django/contrib/auth/__init__.py

```diff
@@ -121,16 +121,22 @@ def login(request, user, backend=None):
     have to reauthenticate on every request. Note that data set during
     the anonymous session is retained when the user logs in.
     """
-    session_auth_hash = ''
+    session_auth_hash_sha256 = ''
+    session_auth_hash_sha1 = ''
     if user is None:
         user = request.user
     if hasattr(user, 'get_session_auth_hash'):
-        session_auth_hash = user.get_session_auth_hash()
```

Comment:
I don't think this will work. The implementation needs to consider the case where a custom user model has implemented its own `get_session_auth_hash()` method (and no `get_session_auth_hash_sha256()` method). I don't have time to think through a solution for this now but the other thing this patches misses as far as I see from a quick glance is tests showing that backwards compatibility is maintained in all the changed places.